llvm/lib/Target/TargetMachine.cpp

```diff
@@ -204,7 +204,8 @@ bool TargetMachine::shouldAssumeDSOLocal(const GlobalValue *GV) const {
     // don't assume the variables to be DSO local unless we actually know
     // that for sure. This only has to be done for variables; for functions
     // the linker can insert thunks for calling functions from another DLL.
-    if (TT.isWindowsGNUEnvironment() && GV->isDeclarationForLinker() &&
+    if ((TT.isWindowsGNUEnvironment() || TT.isWindowsCygwinEnvironment()) &&
```

Comment:
There's a `isOSCygMing()` method in `Triple` which does this `isWindowsCygwinEnvironment() || isWindowsGNUEnvironment()` already.

Also, the subject mentions `[X86]`, but I don't really see anything X86 specific here (even if cygwin mainly is x86-only at the moment), so I'd suggest removing that part.

Reply:
Fixed.